### PR TITLE
Introduce `libnixflake`

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

Flakes is a separate layer, so it should have its own library!

# Context

See each commit for details. The most interesting code changes (that aren't just moving things around) happen in the first commit.

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by edolstra** [comment]
Please keep `registry.*` and `indirect.*` in libfetchers, since they're fetchers and not flake-specific (e.g. you can do `fetchTree` to fetch a repo looked up via the registry). This PR creates a `libflake` that isn't really about flakes at all but about the indirect fetcher, which IMHO belongs in libfetchers.

I prefer to keep all URL/flakeref-related regexes like `flakeIdRegexS` in one place (`url-parts.hh`) rather than scattered all over.

**Comment by Ericson2314** [comment]
```
 $ git grep flake src/libfetchers/
src/libfetchers/fetch-settings.hh:          Example `~/code/flake.nix`:
src/libfetchers/fetch-settings.hh:    Setting<std::string> flakeRegistry{this, "https://channels.nixos.org/flake-registry.json", "flake-registry",
src/libfetchers/fetch-settings.hh:          Path or URI of the global flake registry.
src/libfetchers/fetch-settings.hh:          When empty, disables the global flake registry.
src/libfetchers/fetch-settings.hh:        "Whether to use flake registries to resolve flake references.",
src/libfetchers/fetch-settings.hh:    Setting<bool> acceptFlakeConfig{this, false, "accept-flake-config",
src/libfetchers/fetch-settings.hh:        "Whether to accept nix configuration from a flake without prompting.",
src/libfetchers/fetch-settings.hh:          The commit summary to use when committing changed flake lock files. If
src/libfetchers/fetchers.cc:            // usages (e.g. `builtins.fetchTree` calls or flake inputs).
src/libfetchers/indirect.cc:std::regex flakeRegex("[a-zA-Z][a-zA-Z0-9_-]*", std::regex::ECMAScript);
src/libfetchers/indirect.cc:        if (url.scheme != "flake") return {};
src/libfetchers/indirect.cc:                throw BadURL("in flake URL '%s', '%s' is not a commit hash or branch/tag name", url.url, path[1]);
src/libfetchers/indirect.cc:                throw BadURL("in flake URL '%s', '%s' is not a branch/tag name", url.url, path[1]);
src/libfetchers/indirect.cc:                throw BadURL("in flake URL '%s', '%s' is not a commit hash", url.url, path[2]);
src/libfetchers/indirect.cc:        if (!std::regex_match(id, flakeRegex))
src/libfetchers/indirect.cc:            throw BadURL("'%s' is not a valid flake ID", id);
src/libfetchers/indirect.cc:        if (!std::regex_match(id, flakeRegex))
src/libfetchers/indirect.cc:            throw BadURL("'%s' is not a valid flake ID", id);
src/libfetchers/indirect.cc:        url.scheme = "flake";
src/libfetchers/registry.cc:            for (auto & i : json["flakes"]) {
src/libfetchers/registry.cc:            throw Error("flake registry '%s' has unsupported version %d", path, version);
src/libfetchers/registry.cc:        warn("cannot parse flake registry '%s': %s", path, e.what());
src/libfetchers/registry.cc:        warn("cannot read flake registry '%s': %s", path, e.what());
src/libfetchers/registry.cc:    json["flakes"] = std::move(arr);
src/libfetchers/registry.cc:        auto path = fetchSettings.flakeRegistry.get();
src/libfetchers/registry.cc:            auto storePath = downloadFile(store, path, "flake-registry.json", false).storePath;
src/libfetchers/registry.cc:                store2->addPermRoot(storePath, getCacheDir() + "/nix/flake-registry.json");
src/libfetchers/registry.cc:    if (n > 100) throw Error("cycle detected in flake registry for '%s'", input.to_string());
src/libfetchers/registry.cc:        throw Error("cannot find flake '%s' in the flake registries", input.to_string());
src/libfetchers/tarball.cc:            // FIXME: would be nice to support arbitrary flakerefs
src/libfetchers/tarball.cc:            // here, e.g. git flakes.
```
if we can cut this down I am fine leaving those files in `libfetchers`, but we need to actually make an attempt to *prove* to the user that the feature is not intrinsically part of flakes. Right now I would say it is an "easter egg" the that the **Flake Registry** is not actually about Flakes at all, but "easter egg" simply isn't good enough; it has to be *de jure*.

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nix/pull/9085 My crude attempt to make indirect/registry demonstrably Flake-agnostic by renaming things. I will need some help finishing that PR however.

**Comment by roberth** [comment]
> we shouldn't call it the **Flakes Registry**

This frees up the name for registries in the traditional sense that contain artifacts of sorts.

Still not happy that we overload the term "registry" so I think it's worth taking the opportunity to consider an entirely different name.

The `indirect` scheme is closer to something like DNS than to a registry.

How about **source index**?

**Comment by Ericson2314** [comment]
(The registry stuff is kept in libfetchers now)

File: src/libexpr/eval.cc [hunk 1]
```diff
@@ -21,7 +21,7 @@
 #include "url.hh"
 #include "fetch-to-store.hh"
 #include "tarball.hh"
-#include "flake/flakeref.hh"
+//#include "flake/flakeref.hh"
```
Comment (roberth) [inline src/libexpr/eval.cc:24]:
```suggestion
```

File: src/libflake/flake-settings.hh [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        R"(
+          The commit summary to use when committing changed flake lock files. If
+          empty, the summary is generated based on the action performed.
+        )",
+        {},
+        true,
+        Xp::Flakes};
+};
+
+// FIXME: don't use a global variable.
```
Comment (roberth) [inline src/libflake/flake-settings.hh:50]: It's not great, but is it broken? If not:

```suggestion
// TODO: don't use a global variable.
```

File: src/libexpr/eval-settings.hh [hunk 3]
```diff
@@ -5,9 +5,19 @@
...
 struct EvalSettings : Config
 {
-    EvalSettings(bool & readOnlyMode);
+    /**
+     * Function used to interpet look path entries in the form
+     * `<scheme>:<arbitrary string>`.
+     */
+    using LookupPathHook = std::optional<std::string>(ref<Store> store, std::string_view);
+
+    using LookupPathHooks = std::map<std::string, std::function<LookupPathHook>>;
```
Comment (roberth) [inline src/libexpr/eval-settings.hh:18]:
```suggestion
    /**
     * Map from scheme to LookupPathHook.
     */
    using LookupPathHooks = std::map<std::string, std::function<LookupPathHook>>;
```

File: src/libexpr/eval-settings.hh [hunk 4]
```diff
@@ -5,9 +5,19 @@
...
+class Store;
+
 struct EvalSettings : Config
 {
-    EvalSettings(bool & readOnlyMode);
+    /**
+     * Function used to interpet look path entries in the form
+     * `<scheme>:<arbitrary string>`.
+     */
+    using LookupPathHook = std::optional<std::string>(ref<Store> store, std::string_view);
```
Comment (roberth) [inline src/libexpr/eval-settings.hh:16]: Could you document whether the scheme is passed to the hook, or just the part after it?

Comment (roberth) [reply src/libexpr/eval-settings.hh:16]: Also, what is the meaning of the return value?

**Review by roberth — COMMENTED** [review]
Surprisingly short diff. I like it.
Just a few comments, nothing major.

**Comment by fricklerhandwerk** [comment]
Reminder, this is part of the implementation of RFC 136: https://github.com/NixOS/rfcs/blob/master/rfcs/0136-stabilize-incrementally.md#step-3-attempt-likewise-splitting-a-store-and-language-only-nix

**Review by fricklerhandwerk — APPROVED** [review]

**Review by roberth — APPROVED** [review]

**Comment by cole-h** [comment]
Looks like this accidentally reverted https://github.com/NixOS/nix/pull/10691 (`commit-lock-file-summary` is gone). https://github.com/NixOS/nix/pull/10988 fixes that.

**Comment by Ericson2314** [comment]
Sorry! A negligent rebase conflict resolution on my part, no doubt. Thank you @cole-h for fixing my mistake :heart:.

**Comment by roberth** [comment]
No need to take all the blame; this feature wasn't tested.

**Comment by cole-h** [comment]
Implication heard loud and clear -- I'll add a test in my PR resurrecting it.

**Comment by Ericson2314** [comment]
Tests are def good, and objectively the right way to prevent this, but I also pride myself on being a slicing-and-dicing-like-a-sushi-chef rebase :)r, so I really was dissapointment I slipped up.

**Pull request merged**